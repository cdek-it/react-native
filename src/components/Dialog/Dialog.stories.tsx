import type { Meta, StoryObj } from '@storybook/react'
import { useCallback, useState } from 'react'
import { Text, TouchableOpacity, View } from 'react-native'

import { Dialog } from './Dialog'

const meta: Meta<typeof Dialog> = {
  title: 'Dialog',
  component: Dialog,
  render: () => Template(),
}

const Template = () => {
  const [isVisible, setIsVisible] = useState(false)

  const showDialog = () => {
    setIsVisible(true)
  }

  const hideDialog = () => {
    setIsVisible(false)
  }

  const handleHideComplete = () => {
    console.log('Dialog hide animation completed')
  }

  const Body = useCallback(() => {
    return (
      <Text>Войдите в приложение, чтобы заказ сохранился в списке заказов</Text>
    )
  }, [])

  const Footer = useCallback(() => {
    return (
      <TouchableOpacity onPress={hideDialog}>
        <Text>Войти</Text>
      </TouchableOpacity>
    )
  }, [])

  return (
    <View>
      <View>
        <TouchableOpacity onPress={showDialog}>
          <Text>Показать Dialog</Text>
        </TouchableOpacity>
      </View>

      <Dialog
        body={Body}
        footer={Footer}
        isVisible={isVisible}
        title='Найденный заказ не сохранится'
        onClose={hideDialog}
        onHideComplete={handleHideComplete}
      />
    </View>
  )
}

export default meta

type Story = StoryObj<typeof Dialog>

const DialogStory: Story = { args: {} }

export { DialogStory as Dialog }
