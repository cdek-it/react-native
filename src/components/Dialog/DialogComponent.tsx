import { IconX } from '@tabler/icons-react-native'
import React from 'react'
import { View, TouchableOpacity, useWindowDimensions } from 'react-native'

import { makeStyles } from '../../utils/makeStyles'
import { Title } from '../Typography'

export interface DialogComponentProps {
  readonly title: string
  readonly body?: React.ComponentType
  readonly footer?: React.ComponentType
  readonly onClose?: () => void
}

export const DialogComponent: React.FC<DialogComponentProps> = ({
  title,
  body: Body,
  footer: Footer,
  onClose,
}) => {
  const styles = useStyles()
  const tids = DialogTestIDs
  const { width, height } = useWindowDimensions()
  const maxSize = { maxWidth: width - 40, maxHeight: height - 100 }

  return (
    <View style={[styles.root, maxSize]}>
      <View style={styles.header}>
        <Title level='h3' style={styles.text} testID={tids.title}>
          {title}
        </Title>
        <TouchableOpacity
          style={styles.closeTouchable}
          testID={tids.closeButton}
          onPress={onClose}
        >
          <IconX {...styles.closeIcon} />
        </TouchableOpacity>
      </View>
      <View style={styles.separator} />

      <View style={styles.body}>{Body ? <Body /> : null}</View>

      <View style={styles.footer}>{Footer ? <Footer /> : null}</View>
    </View>
  )
}

const useStyles = makeStyles(({ theme, spacing }) => ({
  root: {
    backgroundColor: theme.Overlay.Dialog.Header.dialogHeaderBg,
    borderColor: theme.Overlay.Overlay.overlayContentBorderColor,
    borderRadius: theme.General.borderRadius,
  },
  body: {
    paddingVertical: theme.Overlay.Dialog.Body.dialogContentPaddingTopBottom,
    paddingHorizontal: theme.Overlay.Dialog.Body.dialogContentPaddingLeftRight,
  },
  footer: {
    paddingBottom: theme.Overlay.Dialog.Footer.dialogFooterPaddingLeftRight,
    paddingHorizontal: theme.Overlay.Dialog.Footer.dialogFooterPaddingLeftRight,
  },
  text: { flexWrap: 'wrap', flex: 1 },
  header: {
    width: '100%',
    gap: spacing.Gap['gap-4'],
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingBottom: theme.Overlay.Dialog.Header.dialogHeaderPaddingTopBottom,
    padding: theme.Overlay.Dialog.Header.dialogHeaderPaddingLeftRight,
  },
  closeTouchable: { padding: 8 },
  separator: {
    height: 1,
    backgroundColor: theme.Overlay.Overlay.overlayContentBorderColor,
  },
  closeIcon: { width: 14, height: 14, color: theme.General.actionIconColor },
}))

const DialogTestIDs = { title: 'DialogTitle', closeButton: 'DialogCloseButton' }
