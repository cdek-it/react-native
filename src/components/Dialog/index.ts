export { Dialog, type DialogProps } from './Dialog'
export { DialogComponent, type DialogComponentProps } from './DialogComponent'
export { DialogHeader, type DialogHeaderProps } from './DialogHeader'
