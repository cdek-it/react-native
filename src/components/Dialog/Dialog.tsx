import React, { useCallback, useEffect, useState } from 'react'
import { View, Modal, StyleSheet } from 'react-native'
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  withSpring,
  runOnJS,
  interpolate,
} from 'react-native-reanimated'

import { DialogComponent, type DialogComponentProps } from './DialogComponent'

const ANIMATION_DURATION = 200
const BACKDROP_OPACITY = 0.5
const SCALE_DAMPING = 5
const SCALE_STIFFNESS = 50
const SCALE_INIT_VALUE = 0.9

export interface DialogProps extends DialogComponentProps {
  readonly isVisible: boolean
  readonly onClose?: () => void
  readonly onHideComplete?: () => void
  readonly testID?: string
}

export const Dialog: React.FC<DialogProps> = ({
  isVisible,
  onClose,
  onHideComplete,
  header,
  footer,
  body,
  testID,
}) => {
  const [modalVisible, setModalVisible] = useState(false)
  const opacity = useSharedValue(0)
  const scale = useSharedValue(SCALE_INIT_VALUE)

  const handleAnimationComplete = useCallback(() => {
    setModalVisible(false)

    if (onHideComplete) {
      onHideComplete()
    }
  }, [onHideComplete])

  useEffect(() => {
    if (isVisible) {
      setModalVisible(true)
      opacity.value = withTiming(1, { duration: ANIMATION_DURATION })
      scale.value = withSpring(1, {
        damping: SCALE_DAMPING,
        stiffness: SCALE_STIFFNESS,
      })
    } else {
      opacity.value = withTiming(0, { duration: ANIMATION_DURATION })
      scale.value = withTiming(
        SCALE_INIT_VALUE,
        { duration: ANIMATION_DURATION },
        (finished) => {
          if (finished && modalVisible) {
            runOnJS(handleAnimationComplete)()
          }
        }
      )
    }
  }, [isVisible, modalVisible, opacity, scale, handleAnimationComplete])

  const backdropAnimatedStyle = useAnimatedStyle(() => {
    return {
      opacity: interpolate(opacity.value, [0, 1], [0, BACKDROP_OPACITY]),
    }
  })

  const dialogAnimatedStyle = useAnimatedStyle(() => {
    return { opacity: opacity.value, transform: [{ scale: scale.value }] }
  })

  return (
    <Modal
      statusBarTranslucent
      transparent
      animationType='none'
      testID={testID ?? DialogTestId.root}
      visible={modalVisible}
      onRequestClose={onClose}
    >
      <View style={styles.container}>
        <Animated.View
          style={[styles.backdrop, backdropAnimatedStyle]}
          testID={DialogTestId.backdrop}
        />
        <Animated.View
          style={dialogAnimatedStyle}
          testID={DialogTestId.contentContainer}
        >
          <DialogComponent body={body} footer={footer} header={header} />
        </Animated.View>
      </View>
    </Modal>
  )
}

const styles = StyleSheet.create({
  container: { flex: 1, justifyContent: 'center', alignItems: 'center' },
  backdrop: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    backgroundColor: 'black',
  },
})

export const DialogTestId = {
  root: 'DialogModal',
  backdrop: 'Backdrop',
  contentContainer: 'ContentContainer',
}
