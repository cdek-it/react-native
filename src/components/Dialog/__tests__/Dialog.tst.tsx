import { render } from '@testing-library/react-native'

import { Text } from 'react-native'

import { Dialog, type DialogProps } from '../Dialog'

describe('Dialog component tests', () => {
  const MockBody = () => <Text>This is the dialog body content</Text>
  const MockFooter = () => <Text>This is the dialog footer</Text>

  const snapshotCases: Array<[string, DialogProps]> = [
    [
      'basic dialog with title, body and footer',
      {
        isVisible: true,
        title: 'Dialog Title',
        body: MockBody,
        footer: MockFooter,
        onClose: jest.fn(),
      },
    ],
  ]

  test.each(snapshotCases)('%s', (_, props) => {
    const renderedDialog = render(<Dialog {...props} />)
    expect(renderedDialog.toJSON()).toMatchSnapshot()
  })
})
