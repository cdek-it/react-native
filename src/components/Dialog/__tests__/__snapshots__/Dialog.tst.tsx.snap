// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog component tests basic dialog with title, body and footer 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  statusBarTranslucent={true}
  transparent={true}
  visible={true}
>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "opacity": 0,
          },
        }
      }
      style={
        {
          "backgroundColor": "black",
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "opacity": 0,
            "transform": [
              {
                "scale": 0.9,
              },
            ],
          },
        }
      }
      style={
        {
          "opacity": 0,
          "transform": [
            {
              "scale": 0.9,
            },
          ],
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "#ffffff",
              "borderColor": "#dbdada",
              "borderRadius": 7,
            },
            {
              "maxHeight": 1234,
              "maxWidth": 710,
            },
          ]
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "gap": 14,
              "justifyContent": "space-between",
              "padding": 21,
              "paddingBottom": 14,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "rgba(0, 0, 0, 0.8000)",
                  "fontWeight": 700,
                  "includeFontPadding": false,
                  "verticalAlign": "middle",
                },
                {
                  "fontSize": 17.5,
                },
                {
                  "flex": 1,
                  "flexWrap": "wrap",
                },
              ]
            }
            testID="DialogTitle"
          >
            Dialog Title
          </Text>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
                "padding": 8,
              }
            }
            testID="DialogCloseButton"
          >
            <RNSVGSvgView
              align="xMidYMid"
              bbHeight={14}
              bbWidth={14}
              fill="none"
              focusable={false}
              height={14}
              meetOrSlice={0}
              minX={0}
              minY={0}
              stroke="rgba(0, 0, 0, 0.6000)"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  {
                    "flex": 0,
                    "height": 14,
                    "width": 14,
                  },
                ]
              }
              vbHeight={24}
              vbWidth={24}
              width={14}
              xmlns="http://www.w3.org/2000/svg"
            >
              <RNSVGGroup
                fill={null}
                propList={
                  [
                    "fill",
                    "stroke",
                    "strokeWidth",
                    "strokeLinecap",
                    "strokeLinejoin",
                  ]
                }
                stroke={
                  {
                    "payload": 2566914048,
                    "type": 0,
                  }
                }
                strokeLinecap={1}
                strokeLinejoin={1}
                strokeWidth={2}
              >
                <RNSVGPath
                  d="M18 6l-12 12"
                  fill={null}
                  propList={
                    [
                      "fill",
                      "stroke",
                      "strokeWidth",
                      "strokeLinecap",
                      "strokeLinejoin",
                    ]
                  }
                  stroke={
                    {
                      "payload": 2566914048,
                      "type": 0,
                    }
                  }
                  strokeLinecap={1}
                  strokeLinejoin={1}
                  strokeWidth={2}
                />
                <RNSVGPath
                  d="M6 6l12 12"
                  fill={null}
                  propList={
                    [
                      "fill",
                      "stroke",
                      "strokeWidth",
                      "strokeLinecap",
                      "strokeLinejoin",
                    ]
                  }
                  stroke={
                    {
                      "payload": 2566914048,
                      "type": 0,
                    }
                  }
                  strokeLinecap={1}
                  strokeLinejoin={1}
                  strokeWidth={2}
                />
              </RNSVGGroup>
            </RNSVGSvgView>
          </View>
        </View>
        <View
          style={
            {
              "backgroundColor": "#dbdada",
              "height": 1,
            }
          }
        />
        <View
          style={
            {
              "paddingHorizontal": 21,
              "paddingVertical": 21,
            }
          }
        >
          <Text>
            This is the dialog body content
          </Text>
        </View>
        <View
          style={
            {
              "paddingBottom": 21,
              "paddingHorizontal": 21,
            }
          }
        >
          <Text>
            This is the dialog footer
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
